cv32e40s/tests/programs/custom/debug_test_trigger/debug_test_trigger.c

```diff
@@ -0,0 +1,1738 @@
+
+/*
+ * execute_test_constructor
+ *
+ * Executes nop.
+ *
+ * Defined as constructor to be placed before main in memory
+ * used for its low address in execute LESS comparisons
+ */
+void execute_test_constructor(void) {
+  __asm__ volatile ("nop");
+}
+
+/*
+ * execute_test_high_addr
+ *
+ * Executes nop.
+ *
+ * Placed in debugger_exception section and used for
+ * its high address in execute test GEQ comparisons
+ *
+ */
+void execute_test_high_addr(void) {
+  __asm__ volatile ("nop");
+}
+
+/*
+ * load_store_test_constructor
+ *
+ * Uncompressed nop, to be used as variable after construction phase.
+ *
+ * Defined as constructor to be placed before main in memory
+ * used for its low address in load/store LESS comparisons
+ */
+void load_store_test_constructor(void) {
+  __asm__ volatile (R"(.option push
+                       .option norvc
+                       nop
+                       .option pop)");
+}
+/*
+ * load_store_test_high_addr
+ *
+ * To be used as variable.
+ *
+ * Placed in debugger_exception section and used for
+ * its high address in load/store test GEQ comparisons
+ */
+void load_store_test_high_addr(void) {
+  __asm__ volatile (".word(0xDEADBEEF)");
+}
+
+/*
+ * handle_illegal_insn
+ *
+ * Sets illegal_insn_status.
+ *
+ * Simple handler used to check illegal intructuction trap
+ */
+void handle_illegal_insn (void) {
+  __asm__ volatile (R"(
+    la   t0,     illegal_insn_status
+    li   t1,     1
+    sw   t1,     0(t0)
+    call end_handler_incr_mepc
+  )");
+}
+
+
+/*
+ * debugger_start
+ *
+ * Debug handler wrapper
+ *
+ * Saves registers, calls debug handler and then restores the registers again.
+ *
+ */
+void _debugger_start(void) {
+  __asm__ volatile (R"(
+    # Store return address and saved registers
+
+      sw a0, -4(sp)
+      sw a1, -8(sp)
+      sw a2, -12(sp)
+      sw a3, -16(sp)
+      sw a4, -20(sp)
+      sw a5, -24(sp)
+      sw a6, -28(sp)
+      sw a7, -32(sp)
+      addi sp, sp, -32
+
+      cm.push {ra, s0-s11}, -64
+
+    # Execute _debug_handler() function
+      call ra, _debug_handler
+
+    # Restore return address and saved registers
+      cm.pop {ra, s0-s11}, 64
+
+      addi sp, sp, 32
+      lw a0, -4(sp)
+      lw a1, -8(sp)
+      lw a2, -12(sp)
+      lw a3, -16(sp)
+      lw a4, -20(sp)
+      lw a5, -24(sp)
+      lw a6, -28(sp)
+      lw a7, -32(sp)
+
+    # Exit debug mode
+      dret
+  )");
+}
+
+/*
+ * _debug_handler
+ *
+ * Debug Handler
+ *
+ * Handles all actions needed in debug mode.
+ *
+ */
+void _debug_handler (void) {
+
+#ifdef DEBUG_PRINT
+  printf("  Entered debug\n");
+#endif
+
+  debug_entry_status = DEBUG_STATUS_ENTERED_OK;
+
+  switch (debug_sel) {
+
+    case DEBUG_SEL_DISABLE_TRIGGER:
+      switch (trigger_type) {
+        case TRIGGER_LOAD_BYTE:
+        case TRIGGER_LOAD_HALFWORD:
+        case TRIGGER_LOAD_WORD:
+          __asm__ volatile ("csrci tdata1, (1 << 0)"); // Clear load bit
+#ifdef DEBUG_PRINT
```

Comment:
I also do not like these as ifdefs; code gets hard to read, and most likely functionally very different

Reply:
I would really like to avoid printfs all together and make some single cycle virtual peripheral print function but I can make it a regular if statement for now.